django/db/models/sql/query.py

```diff
@@ -1156,6 +1156,10 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
         if not arg:
             raise FieldError("Cannot parse keyword query %r" % arg)
         lookups, parts, reffed_expression = self.solve_lookup_type(arg)
+
+        if hasattr(reffed_expression, 'filterable') and not reffed_expression.filterable:
+            raise NotSupportedError(reffed_expression.__class__.__name__ + ' is disallowed in the filter-clause')
+
```

Comment:
There's some overlap with #8119 here. @schinckel what do you think about making `Expression.filterable` a property that returns whether or not `isinstance(self.output_field, BooleanField)` and use it both cases?